Add option to disallow creation of local storages

PVince81 opened the pull request:
Introduce a new config option to prevent web UI admins to create
or edit external storages of type "local".

Fixes https://github.com/nextcloud/server/issues/29973

- [x] TEST: occ command can create
- [x] TEST: cannot create local storages in web UI
- [x] TEST: existing local storages appear grey in web UI
- [x] TEST: test when allowed

~~:warning: this will change the default, by default "Local" will not be visible so one needs to enable it where needed~~

Inspired from:
- https://github.com/owncloud/core/pull/27590
- https://github.com/owncloud/core/pull/27621/files?w=1

PVince81 committed:
Add option to disallow creation of local storages

Introduce a new config option to prevent web UI admins to create
or edit external storages of type "local".

Signed-off-by: Vincent Petry <vincent@nextcloud.com>

PVince81 commented:
I need to fix some tests:
```

1) OCA\Files_External\Tests\Controller\GlobalStoragesControllerTest::testAddStorage
--


ArgumentCountError:
 Too few arguments to function 
OCA\Files_External\Controller\GlobalStoragesController::__construct(), 7
 passed in 
/drone/src/apps/files_external/tests/Controller/GlobalStoragesControllerTest.php
 on line 59 and exactly 8 expected
```

PVince81 commented:
build is passing now, please review

miaulalala commented on apps/files_external/templates/settings.php:
```diff
@@ -100,7 +100,10 @@ function writeParameterInput($parameter, $options, $classes = []) {
 
-<form data-can-create="<?php echo $canCreateMounts?'true':'false' ?>" id="files_external" class="section" data-encryption-enabled="<?php echo $_['encryptionEnabled']?'true': 'false'; ?>">
+<?php
+	$canCreateNewLocalStorage = \OC::$server->getConfig()->getSystemValue('files_external_allow_create_new_local', false);
```
could you use the initial state provider here instead? Pure php in a template gives me the heebies ;)

PVince81 replied:
I'll see what I can do and if it's straightforward

PVince81 replied:
seems that apps/files_external has not be prepped for webpack, ES6 and Vue, so would need to first bootstrap all this before we can make use of initial state and other libs

I think this would blow up the scope of this PR and would rather do so separately.

artonge commented on apps/files_external/js/settings.js:
```diff
@@ -801,6 +803,11 @@ MountConfigListView.prototype = _.extend({
 			};
 		}
 
+		console.log(backend, this._canCreateLocal);
+		if (backend.identifier === 'local' && !this._canCreateLocal) {
+			readOnly = true;
+		}
```
```suggestion
		var readOnly = backend.identifier === 'local' && !this._canCreateLocal
```

PVince81 replied:
I find my version more readable

PVince81 replied:
ah, but I forgot to remove console.log

ok, I'll clean up

artonge commented on apps/files_external/js/settings.js:
```diff
@@ -793,6 +794,7 @@ MountConfigListView.prototype = _.extend({
 	newStorage: function(storageConfig, onCompletion) {
 		var mountPoint = storageConfig.mountPoint;
 		var backend = this._allBackends[storageConfig.backend];
+		var readOnly = false;
```
```suggestion
```

artonge commented on config/config.sample.php:
```diff
@@ -1746,6 +1746,17 @@
+ * It is still possible to create local storages with occ using the following command:
+ * % php occ files_external:create /mountpoint local null::null -c datadir=/path/to/data
+ *
+ * Defaults to ``false``
```
Shouldn't it default to `true` to keep to current behaviour?

szaimen replied:
Yes, I wonder as well. 
I understand the reason but I think for better discoverability it should be true. For instances where this feature is a security problem, it can then be manually disabled.

icewind1991 commented on apps/files_external/js/settings.js:
```diff
@@ -801,6 +803,11 @@ MountConfigListView.prototype = _.extend({
 			};
 		}
 
+		console.log(backend, this._canCreateLocal);
```
left over debug logging :see_no_evil:

PVince81 commented:
fixed the `readOnly` var, I replaced it with an inline statement now as it was used only in one location
and removed console.log and redid a quick test

icewind1991 commented:
/backport to stable23

icewind1991 commented:
/backport to stable22
